internal/installercache/installercache.go

```diff
@@ -183,33 +291,34 @@ func (i *Installers) evict() {
 		return nil
 	}
 
-	err := filepath.Walk(i.cacheDir, visit)
+	err := filepath.Walk(i.config.CacheDir, visit)
 	if err != nil {
 		if !os.IsNotExist(err) { //ignore first invocation where the cacheDir does not exist
-			i.log.WithError(err).Errorf("release binary eviction failed to inspect directory %s", i.cacheDir)
+			i.log.WithError(err).Errorf("release binary eviction failed to inspect directory %s", i.config.CacheDir)
 		}
 		return
 	}
 
-	//prune the hard links just in case the deletion of resources
-	//in ignition.go did not succeeded as expected
-	for idx := 0; idx < len(links); idx++ {
-		finfo := links[idx]
-		//Allow a grace period of 5 minutes from the link creation time
-		//to ensure the link is not being used.
-		grace := time.Now().Add(DeleteGracePeriod).Unix()
-		if finfo.info.ModTime().Unix() < grace {
-			os.Remove(finfo.path)
-		}
-	}
+	// TODO: We might want to consider if we need to do this longer term, in theory every hardlink should be automatically freed.
+	// For now, moved to a function so that this can be tested in unit tests.
+	i.pruneExpiredHardLinks(links, linkPruningGracePeriod)
 
-	//delete the oldest file if necessary
-	for totalSize >= int64(float64(i.storageCapacity)*CacheLimitThreshold) {
-		finfo, _ := files.Pop()
-		totalSize -= finfo.info.Size()
-		//remove the file
-		if err := i.evictFile(finfo.path); err != nil {
-			i.log.WithError(err).Errorf("failed to evict file %s", finfo.path)
+	// delete the oldest file if necessary
+	// prefer files without hardlinks first because
+	// 1: files without hardlinks are likely to be least recently used anyway
+	// 2: files without hardlinks will immediately free up storage
+	queues := i.splitQueueOnHardLinks(files)
```

Comment:
If we're using the actual space used on disk as our metric here why consider evicting files with hardlinks at all?
Removing one of those won't get us any space back.

The amount of extra engineering here is the reason I suggested we use number of releases rather than the actual size for this. This extra complexity is bound to make this fragile, but I understand if we don't want to change directions now.

Reply:
I thought about this and I think there may be cases where we want to trigger eviction sooner.

Consider the scenario where we have capacity for a single release and there are several requests coming in.

In this scenario, the release will always have a hard link.

If we perform an eviction on a release that still has a hard link, this is OK as the actual 'deletion' only happens when all callers are finished with the release.

Reply:
I think there could be limited some edge cases where pruning a release with a hard link could present a small advantage but certainly nothing too significant.

It's safe to include them but in this case, I tend to agree that we should just filter them out.